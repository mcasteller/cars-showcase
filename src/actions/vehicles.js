import database from '../firebase/firebase';

// Actions allow us to change the redux store

// ADD_VEHICLE
export const addVehicle = (vehicle) => ({
  type: 'ADD_VEHICLE',
  vehicle
});

export const startAddVehicle = (vehicleData = {}) => {
  return (dispatch, getState) => {
    const {  
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = vehicleData;
    const vehicle = { description, note, amount, createdAt };

    return database.ref(`vehicles`).push(vehicle).then((ref) => {
      dispatch(addVehicle({
        id: ref.key,
        ...vehicle
      }));
    })
  };
};

// REMOVE_VEHICLE
export const removeVehicle = ({ id } = {}) => ({
  type: 'REMOVE_VEHICLE',
  id
});

export const startRemoveVehicle = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`vehicles/${id}`).remove().then(() => {
      dispatch(removeVehicle({ id }));
    });
  };
};

// EDIT_VEHICLE
export const editVehicle = (id, updates) => ({
  type: 'EDIT_VEHICLE',
  id,
  updates
});

export const startEditVehicle = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`vehicles/${id}`).update(updates).then(() => {
      dispatch(editVehicle(id, updates));
    });
  };
};

// SET_VEHICLES
export const setVehicles = (vehicles) => ({
  type: 'SET_VEHICLES',
  vehicles
});

export const startSetVehicles = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`vehicles`).once('value').then((snapshot) => {
      const vehicles = [];

      snapshot.forEach((childSnapshot) => {
        vehicles.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });

      dispatch(setVehicles(vehicles));
    });
  };
};
