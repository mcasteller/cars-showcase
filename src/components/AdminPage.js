import React from 'react';
import { Link } from 'react-router-dom';

export const AdminPage = (props) => (
  <div className="content-container">
    <Link to="/admin/create">
      <h2>Agregar Vehiculos</h2>
    </Link>
    <Link to="/admin/edit">
      <h2>Editar Vehiculos</h2>
    </Link>
  </div>

);

export default AdminPage;