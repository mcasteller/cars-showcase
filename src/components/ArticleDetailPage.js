import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { connect } from 'react-redux';
import ReactHtmlParser from 'react-html-parser';
import SimpleSlider from './SimpleSlider'; 
import strings from '../resources/strings';

export class ArticleDetailPage extends React.Component {
	constructor(props) {
	    super(props);

	    this.state = {
	      title: props.article ? props.article.title : '', 
	      subTitle: props.article ? props.article.subTitle : '', 
	      description: props.article ? props.article.description : '',
	      filesURL: props.article ? props.article.filesURL : '',
	      error: ''
	    };
	};

	render() {
		return (
		    <div className="content-container">
		      <SimpleSlider
		      	items={this.state.filesURL || [] }
		      	/>
	          <h2>{this.state.title}</h2>
	   	      <h3>{this.state.subTitle}</h3>
		      <span className="vehicle-description">{ReactHtmlParser(this.state.description)}</span>
		    </div>
		)	
	}
}

const mapStateToProps = (state, props) => ({
  article: state.articles.find((article) => article.id === props.match.params.articleId)
});

export default connect(mapStateToProps, undefined)(ArticleDetailPage);