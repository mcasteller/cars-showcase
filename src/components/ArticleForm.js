import React from 'react';
import moment from 'moment';
import ReactQuill from 'react-quill';
import FileSelect from './FileSelect';
import strings from '../resources/strings';

export default class ArticleForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      title: props.article ? props.article.title : '', 
      subTitle: props.article ? props.article.subTitle : '', 
      description: props.article ? props.article.description : '',
      shortDescription: props.article ? props.article.shortDescription : '',
      filesURL: props.article ? props.article.filesURL : '',
      error: ''
    };
  }
  
  onTitleChange = (e) => {
    const title = e.target.value;
    this.setState(() => ({ title }));
  };

  onSubTitleChange = (e) => {
    const subTitle = e.target.value;
    this.setState(() => ({ subTitle }));
  };

  onDescriptionChange = (value) => {
    const description = value;
    this.setState(() => ({ description }));
  };
  
  onShortDescriptionChange = (value) => {
    const shortDescription = value;
    this.setState(() => ({ shortDescription }));
  };
      
  onFilesChange = (filesURL) => {
    this.setState(() => filesURL);
  }

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.title) {
      this.setState(() => ({ error: 'Please provide description and title.' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        title: this.state.title,
        subTitle: this.state.subTitle,
        description: this.state.description,
        shortDescription: this.state.shortDescription,
        filesURL: this.state.filesURL
      });
    }
  };
  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        {this.state.error && <p className="form__error">{this.state.error}</p>}
        <h3>{strings.site.fileRatioWarning}</h3>
        <FileSelect 
          filesURL={this.state.filesURL}
          onChange={this.onFilesChange} 
        />
        <input
          type="textarea"
          placeholder={strings.site.article.title}
          className="text-input"
          value={this.state.title || ''}
          onChange={this.onTitleChange}
        />
        <input
          type="textarea"
          placeholder={strings.site.article.subTitle}
          className="text-input"
          value={this.state.subTitle || ''}
          onChange={this.onSubTitleChange}
        />
        <ReactQuill 
          placeholder={strings.site.article.description}
          value={this.state.description}    
          onChange={this.onDescriptionChange} 
        />
        <ReactQuill
          placeholder={strings.site.article.shortDescription}
          value={this.state.shortDescription}
          onChange={this.onShortDescriptionChange}
        />
        <div>
          <button className="button">{strings.site.article.save}</button>
        </div>
      </form>
    )
  }
}