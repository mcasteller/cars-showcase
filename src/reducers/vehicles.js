const vehiclesReducerDefaultState = [];

// Reducers modify current state based on a given action
// Reducers are PURE functions
// They should return an object that represents the new state
export default (state = vehiclesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_VEHICLE':
      return [
        ...state,
        action.vehicle
      ];
    case 'REMOVE_VEHICLE':
      return state.filter(({ id }) => id !== action.id);
    case 'EDIT_VEHICLE':
      return state.map((vehicle) => {
        if (vehicle.id === action.id) {
          return {
            ...action
          };
        } else {
          return vehicle;
        };
      });
    case 'SET_VEHICLES':
      return action.vehicles;
    default:
      return state;
  }
};
