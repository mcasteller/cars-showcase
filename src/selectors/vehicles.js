import moment from 'moment';
import strings from '../resources/strings';

// Get visible vehicles
export default (vehicles, { text, sortBy}) => {
  return vehicles.filter((vehicle) => {
    const textMatch = vehicle.description.toLowerCase().includes(text.toLowerCase()) ||
                      vehicle.brand.toLowerCase().includes(text.toLowerCase()) ||
                      vehicle.trim.toLowerCase().includes(text.toLowerCase()) ||
                      vehicle.shortDescription.toLowerCase().includes(text.toLowerCase());

    return textMatch;
  }).sort((a, b) => {
    if (sortBy === 'date') {
      return a.year < b.year ? 1 : -1;
    } else if (sortBy === 'amount') {
      return (a.currency === strings.site.currencies.uss ?
              a.amount * 44 : 
              a.amount) < 
             (b.currency === strings.site.currencies.uss ?
              b.amount * 44 : 
              b.amount) 
              ? 1 : -1;
    }
  });
};
