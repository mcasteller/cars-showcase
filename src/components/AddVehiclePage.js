import React from 'react';
import { connect } from 'react-redux';
import VehicleForm from './VehicleForm';
import { startAddVehicle } from '../actions/vehicles';
import strings from '../resources/strings';

class AddVehiclePage extends React.Component {
  onSubmit = (vehicle) => {
    this.props.startAddVehicle(vehicle);
    this.props.history.push('/admin');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">{strings.site.vehicle.addVehicle}</h1>
          </div>
        </div>
        <div className="content-container">
          <VehicleForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

// Make dispatch functions to be available on props 
const mapDispatchToProps = (dispatch) => ({
  startAddVehicle: (vehicle) => dispatch(startAddVehicle(vehicle))
});

// Call higher order component with our component as function call parameter
// Also pass two params to connect component
export default connect(undefined, mapDispatchToProps)(AddVehiclePage);
