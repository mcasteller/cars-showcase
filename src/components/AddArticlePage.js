import React from 'react';
import { connect } from 'react-redux';
import ArticleForm from './ArticleForm';
import { startAddArticle } from '../actions/articles';
import strings from '../resources/strings';

class AddArticlePage extends React.Component {
  onSubmit = (article) => {
    this.props.startAddArticle(article);
    this.props.history.push('/admin');
  };
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">{strings.site.article.addArticle}</h1>
          </div>
        </div>
        <div className="content-container">
          <ArticleForm
            onSubmit={this.onSubmit}
          />
        </div>
      </div>
    );
  }
}

// Make dispatch functions to be available on props 
const mapDispatchToProps = (dispatch) => ({
  startAddArticle: (article) => dispatch(startAddArticle(article))
});

// Call higher order component with our component as function call parameter
// Also pass two params to connect component
export default connect(undefined, mapDispatchToProps)(AddArticlePage);
