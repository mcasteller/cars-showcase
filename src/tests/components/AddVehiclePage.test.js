import React from 'react';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { AddVehiclePage } from '../../components/AddVehiclePage';
import vehicles from '../fixtures/vehicles';

configure({adapter: new Adapter()});

let startAddVehicle, history, wrapper;

beforeEach(() => {
  startAddVehicle = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(<AddVehiclePage startAddVehicle={startAddVehicle} history={history} />);
});

test('should render AddVehiclePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle onSubmit', () => {
  wrapper.find('VehicleForm').prop('onSubmit')(vehicles[1]);
  expect(history.push).toHaveBeenLastCalledWith('/admin');
  expect(startAddVehicle).toHaveBeenLastCalledWith(vehicles[1]);
});
