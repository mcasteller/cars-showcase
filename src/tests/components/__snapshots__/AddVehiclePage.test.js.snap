// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AddVehiclePage correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddVehiclePage
    history={
        Object {
            "push": [Function],
          }
    }
    startAddVehicle={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="page-header"
>
          <div
                    className="content-container"
          >
                    <h1
                              className="page-header__title"
                    >
                              Agregar Vehiculo
                    </h1>
          </div>
</div>,
        <div
          className="content-container"
>
          <VehicleForm
                    onSubmit={[Function]}
          />
</div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="content-container"
>
            <h1
                        className="page-header__title"
            >
                        Agregar Vehiculo
            </h1>
</div>,
          "className": "page-header",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <h1
              className="page-header__title"
>
              Agregar Vehiculo
</h1>,
            "className": "content-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Agregar Vehiculo",
              "className": "page-header__title",
            },
            "ref": null,
            "rendered": "Agregar Vehiculo",
            "type": "h1",
          },
          "type": "div",
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <VehicleForm
            onSubmit={[Function]}
/>,
          "className": "content-container",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSubmit": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="page-header"
>
            <div
                        className="content-container"
            >
                        <h1
                                    className="page-header__title"
                        >
                                    Agregar Vehiculo
                        </h1>
            </div>
</div>,
          <div
            className="content-container"
>
            <VehicleForm
                        onSubmit={[Function]}
            />
</div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="content-container"
>
              <h1
                            className="page-header__title"
              >
                            Agregar Vehiculo
              </h1>
</div>,
            "className": "page-header",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <h1
                className="page-header__title"
>
                Agregar Vehiculo
</h1>,
              "className": "content-container",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Agregar Vehiculo",
                "className": "page-header__title",
              },
              "ref": null,
              "rendered": "Agregar Vehiculo",
              "type": "h1",
            },
            "type": "div",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <VehicleForm
              onSubmit={[Function]}
/>,
            "className": "content-container",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onSubmit": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
